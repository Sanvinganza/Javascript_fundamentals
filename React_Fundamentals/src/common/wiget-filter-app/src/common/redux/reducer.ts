import {
  SELECT_CONTEXT,
  SELECT_DIMENSION,
  SELECT_ITEM,
  SET_STATE
} from './actions.types';

export interface IItems {
    checked: boolean,
    name: string
}

export interface IDimension {
    checked: boolean,
    subcategory: string,
    items: IItems[]
}

export interface IContexts {
  checked: boolean,
    category: string,
    dimensions: IDimension[]
}

export interface IState {
  contexts: IContexts[]
}

const initialState: IState = {
  contexts: []
};

export interface IAction {
    type: string,
    payload: {
        state: IState,
        category: string,
        subcategory: string,
        value: boolean,
        name: string
    }
}

export const rootReducer = (state: IState = initialState, action: IAction) => {
  switch (action.type) {
  case SELECT_ITEM : 
    return {
      contexts: [
        ...state.contexts.map( (context) => {
          return {
            ...context,
            dimensions: [
              ...context.dimensions.map( dimension => {
                return {
                  ...dimension,
                  items: [...dimension.items.map( item => {
                    if(item.name === action.payload.name) {
                      return {
                        ...item,
                        checked: action.payload.value
                      };
                    }

                    return item;
                  })
                  ]
                };
              })
            ]
          };
        })
      ]
    };
  case SELECT_CONTEXT:
    return {
      contexts: [
        ...state.contexts.map( (context) => {
          if(context.category === action.payload.category){
            if(!action.payload.value) {
              return {
                ...context,
                checked: action.payload.value
              };  
            }

            return {
              ...context,
              checked: action.payload.value,
              dimensions: [
                ...context.dimensions.map( 
                  dimension => {
                    return {
                      ...dimension,
                      checked: false
                    };
                  })
              ]
            };
          }
          return context;
        })
      ]
    };
  case SELECT_DIMENSION:
    return {
      contexts: [
        ...state.contexts.map( context => {
          return {
            ...context,
            dimensions: [
              ...context.dimensions.map( dimension => {
                if(dimension.subcategory === action.payload.subcategory){
                  if(!action.payload.value) {
                    return {
                      ...dimension,
                      checked: action.payload.value
                    };
                  }

                  return {
                    ...dimension,
                    checked: action.payload.value,
                    items: [...dimension.items
                      .map( item => {
                        return {
                          ...item,
                          checked: false
                        };
                      })
                    ]
                  };
                }
                
                return dimension;
              })
            ]
          };
        })
      ]
    };
  case SET_STATE:
    return { 
      contexts: [
        ...action.payload.state.contexts
      ]
    };
  default: return state;
  }
};