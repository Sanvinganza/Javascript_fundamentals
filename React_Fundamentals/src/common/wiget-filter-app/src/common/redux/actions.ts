import {
  SELECT_CONTEXT,
  SELECT_DIMENSION,
  SELECT_ITEM,
  SET_STATE
} from './actions.types';
import { IState } from './reducer';

export const selectContext = (category: string, value: boolean) => {
  return {
    type: SELECT_CONTEXT,
    payload: {
      category,
      value
    }
  };
};

export const selectDimension = (subcategory: string, value: boolean) => {
  return {
    type: SELECT_DIMENSION,
    payload: {
      subcategory,
      value
    }
  };
};

export const selectItem = (name: string, value: boolean) => {
  return {
    type: SELECT_ITEM,
    payload: {
      name,
      value
    }
  };
};
  
export const setState = (state: IState) => {
  return {
    type: SET_STATE,
    payload: {
      state
    }
  };
};
  
  