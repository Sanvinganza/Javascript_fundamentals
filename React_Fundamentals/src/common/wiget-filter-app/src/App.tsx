import { Provider } from 'react-redux';
import { WigetFilter } from './common/index';
import { store } from './common/redux/store';
import "react-widgets/scss/styles.scss";

export interface IData {
  [key: string]: {
    [key: string]: Array<string>
  }
}

const filterData: IData  = {
  action: {
    action_section_1: ['action_section_1_Item_3', 'action_section_1_Item_2', 'action_section_1_Item_1'],
    action_section_2: ['action_section_2_Item_1', 'action_section_2_Item_2', 'action_section_2_Item_3'],
    action_section_3: ['action_section_3_Item_1', 'action_section_3_Item_2', 'action_section_3_Item_3']
  },
  comedy: {
    comedy_section_1: ['comedy_section_1_Item_1', 'comedy_section_1_Item_2', 'comedy_section_1_Item_3'],
    comedy_section_2: ['comedy_section_2_Item_1', 'comedy_section_2_Item_2', 'comedy_section_2_Item_3'],
    comedy_section_3: ['comedy_section_3_Item_1', 'comedy_section_3_Item_2', 'comedy_section_3_Item_3']
  },
  drama: {
    drama_section_1: ['drama_section_1_Item_1', 'drama_section_1_Item_2', 'drama_section_1_Item_3'],
    drama_section_2: ['drama_section_2_Item_1', 'drama_section_2_Item_2', 'drama_section_2_Item_3'],
    drama_section_3: ['drama_section_3_Item_1', 'drama_section_3_Item_2', 'drama_section_3_Item_3']
  },
  fantasy: {
    fantasy_section_1: ['fantasy_section_1_Item_1', 'fantasy_section_1_Item_2', 'fantasy_section_1_Item_3'],
    fantasy_section_2: ['fantasy_section_2_Item_1', 'fantasy_section_2_Item_2', 'fantasy_section_2_Item_3'],
    fantasy_section_3: ['fantasy_section_3_Item_1', 'fantasy_section_3_Item_2', 'fantasy_section_3_Item_3']
  },
  horror: {
    horror_section_1: ['horror_section_1_Item_1', 'horror_section_1_Item_2', 'horror_section_1_Item_3'],
    horror_section_2: ['horror_section_2_Item_1', 'horror_section_2_Item_2', 'horror_section_2_Item_3'],
    horror_section_3: ['horror_section_3_Item_1', 'horror_section_3_Item_2', 'horror_section_3_Item_3']
  },
};

export default function App() {
  return (
    <Provider store={store}>
      <WigetFilter data={filterData} />
    </Provider>
  );
}