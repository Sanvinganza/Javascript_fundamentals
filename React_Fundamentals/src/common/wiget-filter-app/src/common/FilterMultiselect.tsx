import { useCallback } from "react";
import { useDispatch } from "react-redux";
import { Multiselect } from "react-widgets/cjs";
import { IItem } from "./Contexts";
import { IState } from "./redux/reducer";

interface IFilterMultiselectProps {
  data?: string[],
  value?: string[],
  showPlaceholderWithValues?: boolean,
  showSelectedItemsInList?: boolean,
  filter?: string,
  open?: boolean,
  Item: ({item}: IItem) => JSX.Element,
  selectedItems?: string[],
  selectAction: (subcategory: string, value: boolean) =>  {
    type: string,
    payload: {
        state?: IState,
        category?: string,
        subcategory?: string,
        value?: boolean,
        name?: string
    }
  }
}

const arraySelectedItems: Array<string> = [];

export const FilterMultiselect = (
  {data, value, showPlaceholderWithValues, filter, 
    showSelectedItemsInList, open, Item, selectAction}: IFilterMultiselectProps
) => {
  const dispatch = useDispatch();

  return <Multiselect
    onSelect={useCallback((item: string) => {
      if(arraySelectedItems.includes(item)) {
        dispatch(selectAction(item, false));
        arraySelectedItems.splice(arraySelectedItems.indexOf(item), 1);
      } else {
        dispatch(selectAction(item, true));
        arraySelectedItems.push(item);
      }
    }, [data])}
    value={value}
    filter={filter}
    showPlaceholderWithValues={showPlaceholderWithValues}
    showSelectedItemsInList={showSelectedItemsInList}
    data={data}
    renderListItem={({item}: IItem) => <Item item={item}/>}
    open={open}
  />;
};