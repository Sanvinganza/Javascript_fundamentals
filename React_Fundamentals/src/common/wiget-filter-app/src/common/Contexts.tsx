import { Checkbox } from 'antd';
import { useSelector } from 'react-redux';
import { Dismensions } from './Dimensions';
import { getSelectedContextsSelector } from './selectors/Contexts/getSelectedContextsSelector';
import { getContextsSelector } from './selectors/Contexts/getContextsSelector';
import { getIsContextSelectedSelector } from './selectors/Contexts/getIsContextSelectedSelector';
import { useMemo } from 'react';
import { selectContext } from './redux/actions';
import { IContexts } from './redux/reducer';
import { FilterMultiselect } from './FilterMultiselect';

export interface IItem {
  item: string
}

const CheckboxContext = ({item}: IItem) => {
  const checked = useSelector(getIsContextSelectedSelector(item));

  return <Checkbox checked={checked}>{item}</Checkbox>;
};

export function Contexts () {

  const contexts = useSelector(getContextsSelector());
  const memoContexts = useMemo(() => contexts.map((item: IContexts) => item.category), contexts);
  
  const selectedContexts = useSelector(getSelectedContextsSelector());

  return (
    <>
      <h2>CONTEXTS</h2> 
      <FilterMultiselect 
        showPlaceholderWithValues={true}
        showSelectedItemsInList={true}
        data={memoContexts}
        Item={CheckboxContext}
        selectAction={selectContext}
      />
      {!!selectedContexts.length && <Dismensions />}
    </>
  );
}
