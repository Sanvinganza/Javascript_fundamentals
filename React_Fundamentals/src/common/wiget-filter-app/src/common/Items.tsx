import { useSelector } from "react-redux";
import { Combobox } from "react-widgets/cjs";
import { getItemsSelector } from "./selectors/Items/getItemsSelector";
import { getSelectedItemsSelector } from "./selectors/Items/getSelectedItemsSelector";
import { useState } from "react";
import { getReverseSortedDataSelector } from "./selectors/Items/getReverseSortedDataSelector";
import { getSortedDataSelector } from "./selectors/Items/getSortedDataSelector";
import { SortButton, SortingMode } from "./SortButton";
import { IItems } from "./redux/reducer";
import { FilterMultiselect } from "./FilterMultiselect";
import { IItem } from "./Contexts";
import { Checkbox } from "antd";
import { getIsItemSelectedSelector } from "./selectors/Items/getIsItemSelectedSelector";
import { selectItem } from "./redux/actions";

export function Items () {
  const selectedItems = useSelector(getSelectedItemsSelector()).map((selectedItem: IItems) => selectedItem.name);
  
  const data = useSelector(getItemsSelector());
  const reverseSortedData = useSelector(getReverseSortedDataSelector());
  const sortedData = useSelector(getSortedDataSelector());
  
  const [selectFilter, setSelectFilter] = useState('startsWith');
  const [mode, setMode] = useState(SortingMode.unsorted);

  const Item = ({item}: IItem) => {
    const checked = useSelector(getIsItemSelectedSelector(item));
    return <Checkbox checked={checked}>{item}</Checkbox>;
  };
  
  return (
    <>
      <div className="optionsItems">
        <SortButton mode={mode} setMode={setMode}/>
        <Combobox
          data={['**', '*_*_', '*_']}
          defaultValue={'*_'}
          onSelect={(item: string) => {
            switch(item) {
            case '*_': setSelectFilter('startsWith'); break;
            case '**': setSelectFilter('eq'); break;
            case '*_*_': setSelectFilter('contains'); break;
            }
          }}
        />
      </div>
      <div className="items-container">
        <FilterMultiselect 
          showPlaceholderWithValues={true}
          showSelectedItemsInList={true}
          data={mode === 'reverseSorted'?
            reverseSortedData: mode === 'sorted'?
              sortedData: data}
          open={true}
          filter={selectFilter}
          value={selectedItems}
          Item={Item}
          selectAction={selectItem}
        />
      </div>
    </>
  );
}