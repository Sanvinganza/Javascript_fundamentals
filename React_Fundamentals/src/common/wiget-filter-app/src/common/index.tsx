import { Button } from 'antd';
import * as React from 'react';
import 'antd/dist/antd.css';
import { useDispatch } from 'react-redux';
import { Contexts } from './Contexts';
import { IData } from '../App';
import { setState } from './redux/actions';
import { DraggableModal, DraggableModalProvider } from 'ant-design-draggable-modal';

export interface IWigetFilterProps  {
  data: IData
}

export const dataToState = (data: IData) => {
  return {
    contexts: Object.entries(data).map( context => {
      return {
        category: context[0],
        checked: false,
        dimensions: Object.entries(context[1]).map( dimensions => {
          return {
            subcategory: dimensions[0],
            checked: false,
            items: dimensions[1].map( item => {
              return {
                checked: false,
                name: item
              };
            })};
        })};
    })
  };
};

export function WigetFilter ({data}: IWigetFilterProps) {
  const [isModalVisible, setIsModalVisible] = React.useState(false);
  const dispatch = useDispatch();
  const state = dataToState(data);

  const showModal = () => {
    dispatch(setState(state));
    setIsModalVisible(true);
  };
  
  const handleCancel = () => {
    setIsModalVisible(false);
  };

  return (
    <DraggableModalProvider>
      <Button type="primary" onClick={showModal}>
          Open Filter
      </Button>
      <DraggableModal
        footer={false} 
        title="FILTERS" 
        visible={isModalVisible} 
        onCancel={handleCancel}
      >
        <Contexts />
      </DraggableModal>
    </DraggableModalProvider>
  );
}
