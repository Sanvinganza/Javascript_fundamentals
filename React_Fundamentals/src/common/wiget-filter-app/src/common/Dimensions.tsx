import { Checkbox } from 'antd';
import { useSelector } from 'react-redux';
import { IItem } from './Contexts';
import { Items } from './Items';
import { getDimensionsSelector } from './selectors/Dimensions/getDimensionsSelector';
import { getIsDimensionSelectedSelector } from './selectors/Dimensions/getIsDimensionSelectedSelector';
import { useMemo } from 'react';
import { selectDimension } from './redux/actions';
import { IDimension } from './redux/reducer';
import { FilterMultiselect } from './FilterMultiselect';

const CheckboxDimension = ({item}: IItem) => {
  const checked = useSelector(getIsDimensionSelectedSelector(item));
  return <Checkbox checked={checked}>{item}</Checkbox>;
};

export function Dismensions () {
  const dimensions = getDimensionsSelector();
  const selectedDimensions = dimensions.filter(dimension => dimension.checked);
  const memoDimensions = useMemo(() => dimensions.map((dimension: IDimension) => dimension.subcategory), [dimensions]);

  return (
    <>
      <h2>DISMENSIONS</h2> 
      <FilterMultiselect
        value={selectedDimensions.map((selectedDimension: IDimension) => selectedDimension.subcategory)}
        showSelectedItemsInList={true}
        data={memoDimensions}
        Item={CheckboxDimension}
        showPlaceholderWithValues={true}
        selectAction={selectDimension}
        selectedItems={selectedDimensions.map((selectedDimension: IDimension) => selectedDimension.subcategory)}
      />
      {!!selectedDimensions.length && <Items />}
    </>
  );
}
