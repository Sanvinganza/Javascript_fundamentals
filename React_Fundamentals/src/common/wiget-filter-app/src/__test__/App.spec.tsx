import { configureStore } from "@reduxjs/toolkit";
import { render, fireEvent, RenderOptions } from "@testing-library/react"
import { Provider } from "react-redux";
import App from "../App"
import { Contexts } from "../common/Contexts";
import { store } from "../common/redux/store";
import { PreloadedState } from 'redux';
import { initialState, IState, rootReducer } from "../common/redux/reducer";
import { PropsWithChildren } from "react";

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  store?: typeof store
}

function renderWithProviders(
  ui: React.ReactElement,
  {
    store =  configureStore({reducer: rootReducer})
  }: ExtendedRenderOptions = {}
) {

  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  
  return { store, ...render(ui, { wrapper: Wrapper }) }
}

describe('App component', () => {
  it('is renderer', () => {
    expect(renderWithProviders(<App />)).toMatchSnapshot();
  })
  it('should be show WigetFilter after pressed button', async () => {
    const app = render(<App />);

    const button = app.getByRole('button');
    
    await fireEvent.click(button);

    await expect(render(
        <Contexts />
    )).toMatchSnapshot();
  })
})