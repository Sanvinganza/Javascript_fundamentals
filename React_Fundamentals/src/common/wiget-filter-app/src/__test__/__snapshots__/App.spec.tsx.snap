// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component is renderer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="ant-btn ant-btn-primary"
        type="button"
      >
        <span>
          Open Filter
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="ant-btn ant-btn-primary"
      type="button"
    >
      <span>
        Open Filter
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App component should be show WigetFilter after pressed button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="ant-scrolling-effect"
    style="overflow: hidden; overflow-x: hidden; overflow-y: hidden;"
  >
    <div>
      <button
        ant-click-animating-without-extra-node="false"
        class="ant-btn ant-btn-primary"
        type="button"
      >
        <span>
          Open Filter
        </span>
      </button>
    </div>
    <div>
      <div
        class="ant-modal-root"
      >
        <div
          class="ant-modal-wrap ant-design-draggable-modal"
          style="z-index: 2;"
          tabindex="-1"
        >
          <div
            aria-labelledby="test-id"
            aria-modal="true"
            class="ant-modal"
            role="dialog"
            style="margin: 0px; padding-bottom: 0px; pointer-events: auto; top: 0px; left: 112px; height: 768px; width: 800px;"
          >
            <div
              aria-hidden="true"
              style="width: 0px; height: 0px; overflow: hidden; outline: none;"
              tabindex="0"
            />
            <div
              class="ant-modal-content"
            >
              <button
                aria-label="Close"
                class="ant-modal-close"
                type="button"
              >
                <span
                  class="ant-modal-close-x"
                >
                  <span
                    aria-label="close"
                    class="anticon anticon-close ant-modal-close-icon"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
              <div
                class="ant-modal-header"
              >
                <div
                  class="ant-modal-title"
                  id="test-id"
                >
                  <div
                    class="ant-design-draggable-modal-title"
                  >
                    FILTERS
                  </div>
                </div>
              </div>
              <div
                class="ant-modal-body"
              >
                <h2>
                  CONTEXTS
                </h2>
                <div
                  class="rw-multiselect rw-widget"
                  data-intent="mouse"
                  tabindex="-1"
                >
                  <div
                    class="rw-widget-input rw-widget-picker rw-widget-container"
                  >
                    <div
                      aria-label="Selected items"
                      aria-multiselectable="true"
                      aria-orientation="horizontal"
                      class="rw-multiselect-taglist"
                      id="rw_2_taglist"
                      role="listbox"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-busy="false"
                        aria-controls="rw_3_listbox"
                        aria-disabled="false"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-owns="rw_3_listbox "
                        aria-readonly="false"
                        autocapitalize="off"
                        autocomplete="off"
                        class="rw-multiselect-input"
                        placeholder=""
                        role="combobox"
                        size="2"
                        spellcheck="false"
                        tabindex="0"
                        value=""
                      />
                    </div>
                    <span
                      aria-hidden="true"
                      class="rw-btn rw-picker-caret"
                    />
                  </div>
                </div>
                <div
                  class="ant-design-draggable-modal-resize-handle"
                >
                  <div
                    class="ant-design-draggable-modal-resize-handle-inner"
                  />
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              style="width: 0px; height: 0px; overflow: hidden; outline: none;"
              tabindex="0"
            />
          </div>
        </div>
      </div>
    </div>
    <div>
      <h2>
        CONTEXTS
      </h2>
      <div
        class="rw-multiselect rw-widget"
        data-intent="mouse"
        tabindex="-1"
      >
        <div
          class="rw-widget-input rw-widget-picker rw-widget-container"
        >
          <div
            aria-label="Selected items"
            aria-multiselectable="true"
            aria-orientation="horizontal"
            class="rw-multiselect-taglist"
            id="rw_8_taglist"
            role="listbox"
          >
            <input
              aria-autocomplete="list"
              aria-busy="false"
              aria-controls="rw_9_listbox"
              aria-disabled="false"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-owns="rw_9_listbox "
              aria-readonly="false"
              autocapitalize="off"
              autocomplete="off"
              class="rw-multiselect-input"
              placeholder=""
              role="combobox"
              size="2"
              spellcheck="false"
              tabindex="0"
              value=""
            />
          </div>
          <span
            aria-hidden="true"
            class="rw-btn rw-picker-caret"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h2>
      CONTEXTS
    </h2>
    <div
      class="rw-multiselect rw-widget"
      data-intent="mouse"
      tabindex="-1"
    >
      <div
        class="rw-widget-input rw-widget-picker rw-widget-container"
      >
        <div
          aria-label="Selected items"
          aria-multiselectable="true"
          aria-orientation="horizontal"
          class="rw-multiselect-taglist"
          id="rw_8_taglist"
          role="listbox"
        >
          <input
            aria-autocomplete="list"
            aria-busy="false"
            aria-controls="rw_9_listbox"
            aria-disabled="false"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="rw_9_listbox "
            aria-readonly="false"
            autocapitalize="off"
            autocomplete="off"
            class="rw-multiselect-input"
            placeholder=""
            role="combobox"
            size="2"
            spellcheck="false"
            tabindex="0"
            value=""
          />
        </div>
        <span
          aria-hidden="true"
          class="rw-btn rw-picker-caret"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
